solver/llbsolver/vertex.go

```diff
@@ -401,6 +326,117 @@ func fileOpName(actions []*pb.FileAction) string {
-	return proto.MarshalOptions{Deterministic: true}.Marshal(m)
+type opMutator func(context.Context, *pb.Op) error
+
+func recomputeDigests(ctx context.Context, def *pb.Definition, mutators ...opMutator) (*pb.Definition, *digestMapping, error) {
+	dm, err := newDigestMapping(def)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := dm.Rewrite(ctx, mutators); err != nil {
+		return nil, nil, err
+	}
+	return dm.out, dm, nil
+}
+
+type digestMapping struct {
+	in, out       *pb.Definition
+	mapping       map[digest.Digest]digest.Digest
+	indexByDigest map[digest.Digest]int
+	opByDigest    map[digest.Digest]*pb.Op
+}
+
+func newDigestMapping(def *pb.Definition) (*digestMapping, error) {
+	dm := &digestMapping{
+		in:            def,
+		out:           def.CloneVT(),
+		mapping:       map[digest.Digest]digest.Digest{},
+		indexByDigest: map[digest.Digest]int{},
+		opByDigest:    map[digest.Digest]*pb.Op{},
+	}
+	for i, in := range def.Def {
+		dgst := digest.FromBytes(in)
+
+		op := new(pb.Op)
+		if err := op.UnmarshalVT(in); err != nil {
+			return nil, errors.Wrap(err, "failed to parse llb proto op")
+		}
+		dm.opByDigest[dgst] = op
+		dm.indexByDigest[dgst] = i
+	}
+	return dm, nil
+}
+
+func (dm *digestMapping) Rewrite(ctx context.Context, mutators []opMutator) error {
+	for dgst := range dm.indexByDigest {
+		if _, err := dm.rewrite(ctx, dgst, mutators); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (dm *digestMapping) OpFor(dgst digest.Digest) *pb.Op {
+	return dm.opByDigest[dgst]
+}
+
+func (dm *digestMapping) IndexFor(dgst digest.Digest) int {
+	index, ok := dm.indexByDigest[dgst]
+	if !ok {
+		return -1
+	}
+	return index
+}
+
+func (dm *digestMapping) rewrite(ctx context.Context, dgst digest.Digest, mutators []opMutator) (digest.Digest, error) {
+	if dgst, ok := dm.mapping[dgst]; ok {
+		return dgst, nil
+	}
+
+	op, ok := dm.opByDigest[dgst]
+	if !ok {
+		return "", errors.Errorf("invalid missing input digest %s", dgst)
+	}
+
+	for _, mutator := range mutators {
+		if err := mutator(ctx, op); err != nil {
+			return "", err
+		}
+	}
+
+	// Recompute input digests.
+	for _, input := range op.Inputs {
+		select {
+		case <-ctx.Done():
+			return "", context.Cause(ctx)
+		default:
+		}
+
+		iDgst, err := dm.rewrite(ctx, digest.Digest(input.Digest), mutators)
+		if err != nil {
+			return "", err
+		}
+		if digest.Digest(input.Digest) != iDgst {
+			input.Digest = string(iDgst)
+		}
+	}
+
+	// Must use deterministic marshal here so the digest is consistent.
+	data, err := op.Marshal()
+	if err != nil {
+		return "", err
+	}
+	newDgst := digest.FromBytes(data)
+
+	dm.mapping[dgst] = newDgst
+	if dgst != newDgst {
+		// Ensure the indices also map to the new digest.
+		dm.indexByDigest[newDgst] = dm.indexByDigest[dgst]
+		dm.opByDigest[newDgst] = dm.opByDigest[dgst]
+		dm.mapping[newDgst] = newDgst
+	}
+
+	index := dm.indexByDigest[dgst]
+	dm.out.Def[index] = data
```

Comment:
Do we need this? We have already parsed the data once. If we write back the encoded data that means we need to parse it again. We should only need to marshal to get the new digest, don't actually need the new bytes.

Reply:
Hm I'll go back and see if this is an easier way of doing it. I think you're likely right. I'll give it a try.